Allowed named classes and functions as BlockNames

JoshuaKGoldberg opened the pull request:
**Summary**

`describe` can now take in classes and functions as the names. See #5152

**Test plan**

I'm curious about this actually - where would you like me to add tests? It wasn't very clear scanning through the code.

_(first Jest PR; still getting used to Flow... expect failures the first few commits!)_

JoshuaKGoldberg commented on packages/jest-jasmine2/src/jasmine/Suite.js:
```diff
@@ -181,6 +181,25 @@ Suite.prototype.addExpectationResult = function() {
+  }
+
+  const stringified = descriptor.toString();
+  const typeDescriptorMatch = stringified.match(/class|function/);
```
If something silly, such as a number or `{ toString() { return "nope"; }` is passed here, it'll crash. How would you like this function to deal with those error handling cases?

cpojer replied:
Can you just make the function throw in that case and ask people to provide a string, class or function instead?

SimenB commented:
Thanks for the PR!

I'm curios, is this needed? Why can't you do `describe(Adder.toString(), () => {})`?

cpojer commented:
I agree with @SimenB. You can get all the benefits of easy refactoring by doing this manually without having to change Jest.

JoshuaKGoldberg commented:
`Adder.toString()` does not actually work well - it fills up the terminal with the full contents of the class/function.

Using `create-react-app`'s default setup, and adding a `describe(Adder` or `describe(Adder.toString()` to `App.test.js`, we get the lovely:

```
 PASS  src\App.test.js
  class App extends _react.Component {
  render() {
    return (
      _react2.default.createElement('div', { className: 'App', __source: { fileName: _jsxFileName, lineNumber: 8 }, __self: this },
        _react2.default.createElement('header', { className: 'App-header', __source: { fileName: _jsxFileName, lineNumber: 9 }, __self: this },
          _react2.default.createElement('img', { src: _logo2.default, className: 'App-logo', alt: 'logo', __source: { fileName: _jsxFileName, lineNumber: 10 }, __self: this }),
          _react2.default.createElement('h1', { className: 'App-title', __source: { fileName: _jsxFileName, lineNumber: 11 }, __self: this }, 'Welcome to React')),

        _react2.default.createElement('p', { className: 'App-intro', __source: { fileName: _jsxFileName, lineNumber: 13 }, __self: this }, 'To get started, edit ',
          _react2.default.createElement('code', { __source: { fileName: _jsxFileName, lineNumber: 14 }, __self: this }, 'src/App.js'), ' and save to reload.')));



  }}
    √ renders without crashing (17ms)
```

@cpojer could you please re-open the PR & issue?

SimenB commented:
What about `Adder.name`?

JoshuaKGoldberg commented:
Undefined in IE. Two reasons why this is relevant:
* We're talking about letting Jest run in-browser, a common use case for which is old browser compat testing
* Some teams that are on TypeScript use modified versions of the standard `lib.d.ts` that exclude properties not known to work in all browsers, such as `Function::name`. That means they can't be used in tests either. Presumably Flow folks could have a similar situation.

SimenB commented:
Hmm, OK. You can override `describe` in http://facebook.github.io/jest/docs/en/configuration.html#setuptestframeworkscriptfile-string, but maybe not as elegant

cpojer commented:
Alright, I thought about this some more and have no reason to push back on it. Please add an integration test for this feature before we can merge it.

SimenB commented:
This is failing on all CIs, something with the snapshot that was created is wrong (missing a number). It's also weird that you only have a single snapshot as you call `toMatchSnapshot` twice

SimenB commented on CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ None for now
 
 ### Chore & Maintenance
 
+## jest 22.0.5
```
leave this out, it says "master" at the top 🙂

JoshuaKGoldberg commented:
Added a thrown error if what's passed to describe isn't a string, class, or function. I tried adding a snapshot & test case (https://github.com/facebook/jest/pull/5154/commits/c4b83438c33f6c6d22eef3336192b218e70c485d) but CI required source lines including `..\circle-ci`, and I couldn't find an example of existing code that has a similar test case type. Removed in https://github.com/facebook/jest/pull/5154/commits/e1f6c7fdc7773081fb0d5fe0a443169b7d1edfb6.

Also: I show edthis PR off-hand to a friend and they mentioned they intentionally use numbers as their `describe` descriptors in some tests. Added numbers as an allowed `typeof`. I'm +0 to only doing the `.name` logic for function types, and otherwise just returning as before for others (to more closely mirror the existing behavior)... will change the PR if asked to. :)

cpojer commented:
Thanks for making this Pr :)

SimenB commented on packages/jest-jasmine2/src/jasmine/Suite.js:
```diff
@@ -181,6 +181,33 @@ Suite.prototype.addExpectationResult = function() {
+  const typeDescriptorMatch = stringified.match(/class|function/);
+  const indexOfNameSpace =
+    typeDescriptorMatch.index + typeDescriptorMatch[0].length;
+  const indexOfNameAfterSpace = stringified.search(/\(|\{/, indexOfNameSpace);
```
@JoshuaKGoldberg I'm looking at converting this to TS, and it's not happy about the second argument here. It's also not documented at MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/search

Should I just remove it?

JoshuaKGoldberg replied:
@SimenB ha, sure.

Looks like yet another "bug" fixed by TypeScript! 😍
